[11.0][ADD] project_task_pull_request_state

InsanivimusOmnes opened the pull request:
This module extends functionality of the project_task_pull_request module.
It adds a "State" field to Task alongside with PR URI field.
Following pre-defined states are available: "Draft", "Open", "Merged", "Closed". Default state is "Open".
It is possible to filter or group tasks based on their state.

OCA-git-bot committed:
[ADD] project_task_pull_request_state: add a "State" field to Task alongside with PR URI field

ivs-cetmix commented on project_task_pull_request/__manifest__.py:
```diff
@@ -2,17 +2,18 @@
     "category": "Project Management",
     "website": "https://odoo-community.org/",
-    "author": "SMDrugstore, Odoo Community Association (OCA)",
+    "author": "SMDrugstore, Cetmix, Odoo Community Association (OCA)",
```
Remove Cetmix because it is not the module author

ivs-cetmix commented on project_task_pull_request/__manifest__.py:
```diff
@@ -2,17 +2,18 @@
     "depends": [
         "project",
     ],
+    "maintainers": ["InsanivimusOmnes"],
```
This is an existing module. Maintainers must not be updated

ivs-cetmix commented on project_task_pull_request_state/models/project_task_state.py:
```diff
@@ -0,0 +1,15 @@
+        ('draft', 'Draft'),
+        ('merged', 'Merged'),
+        ('closed', 'Closed'),
+    ], string="PR State", default='open', required=True)
```
Why is this field required?

ivs-cetmix reviewed (changes requested):
PR must contain a single module only.
No need to add tests in case there is nothing to be tested.

ivs-cetmix commented on project_task_pull_request_state/models/project_task_state.py:
```diff
@@ -0,0 +1,15 @@
+        ('draft', 'Draft'),
+        ('merged', 'Merged'),
+        ('closed', 'Closed'),
+    ], string="PR State", default='open')
```
Remove "default" for now. PR state must be unset on task creation because might be no PR at all.

ivs-cetmix commented on project_task_pull_request/tests/test_project_task_pull_request.py:
```diff
@@ -57,3 +57,12 @@ def test_write_allowed_with_pr(self):
         self.task_3.refresh()
         self.assertEquals(6, self.task_3.stage_id.id)
+
+    def test_pr_uri(self):
```
This test is pointless because it is covering standard "write"->"read" behaviour. Not the current module flow.
So you should remove it.

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="model">project.task</field>
+        <field name="inherit_id" ref="project.view_task_form2"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='user_id']" position="after">
```
Why are you adding this field here?

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="model">project.task</field>
+        <field name="inherit_id" ref="project.view_task_form2"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='tag_ids']" position="after">
```
Put this field right after the "pr_uri" field.

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -0,0 +1,24 @@
+            </xpath>
+        </field>
+    </record>
+    <record id="project_task_pr_state" model="ir.ui.view">
```
Inherti view from the module your module depends on

ivs-cetmix replied:
This is not solved

ivs-cetmix commented on project_task_pull_request_state/models/project_task_state.py:
```diff
@@ -0,0 +1,15 @@
+class ProjectTaskState(models.Model):
+    _inherit = "project.task"
+
+    pr_state = fields.Selection([
```
Better use function to define variables because it will be easier to override values later if needed.
See example[ here](https://github.com/OCA/stock-logistics-warehouse/blob/5f576f26eaa110a7f015c3c14a135a2fb185e846/stock_location_product_restriction/models/stock_location.py#L13)

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="project_view_task_form_pr_state" model="ir.ui.view">
```
Record id of the inheriting view should be the same as the inherited one.
See [OCA guidlines](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#253inherited-xml)

ivs-cetmix reviewed (changes requested):
PR must contain new module only.
Please set PR to draft

ivs-cetmix commented on project_task_pull_request_state/__manifest__.py:
```diff
@@ -10,9 +8,12 @@
     "installable": True,
-    "depends": ["project_task_pull_request"],
+    "depends": [
+        "project",
```
project_task_pull_request already depends on project. No need to add it explicitly

ivs-cetmix commented on project_task_pull_request_state/__manifest__.py:
```diff
@@ -10,9 +8,12 @@
+        "project",
+        "project_task_pull_request",
+    ],
     "maintainers": ["InsanivimusOmnes"],
```
add CetmixGitDrone to maintainers

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -12,4 +12,18 @@
+        <field name="inherit_id" ref="project.view_task_search_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='pr_uri']" position="after">
+                <field name="pr_state" domain="[('pr_state', 'ilike', self)]"/>
```
Why are you adding "domain=" here?

ivs-cetmix commented on project_task_pull_request_state/models/project_project.py:
```diff
@@ -0,0 +1,18 @@
+        selection=lambda self: self._selection_pr_state(), string="PR State default")
+
+    @api.model
+    def _selection_pr_state(self):
```
Use existing method from project.task

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,31 @@
+
+    @api.multi
+    def write(self, vals):
+        if 'pr_uri' in vals:
```
Put this logic into a separate function and call it from write

ivs-cetmix reviewed (changes requested):
1. Check comments
2. You need to check PR state also on task creation

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,39 @@
+
+    @api.model
+    def create(self, vals):
+        self._check_pr_state(vals)
```
_check_pr_state() function requires record set in self.
However there is are no records at this point because nothing is created yet

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,42 @@
+
+    @api.model
+    def create(self, vals):
+        if 'pr_uri' in self:
```
how "pr_uri" can appear in "self"?

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,42 @@
+    def create(self, vals):
+        if 'pr_uri' in self:
+            search_pr_state = self.env['project.project'].search(
+                [('id', '=', vals.get('project_id'))]).pr_state_default
```
What if there is no "project_id" in "vale"?

ivs-cetmix commented on project_task_pull_request_state/views/project_task_pull_request_state_view.xml:
```diff
@@ -0,0 +1,37 @@
+        <field name="inherit_id" ref="project.view_task_search_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='stage_id']" position="after">
+                <field name="pr_state"/>
```
"pr_state" should be visible only if there is a value in "pr_uri"

ivs-cetmix reviewed (changes requested):
1. Add tests for "create" method.
2. Use "_check_pr_state" function also in create

geomer198 commented on project_task_pull_request_state/models/project_project.py:
```diff
@@ -0,0 +1,10 @@
+    _name = "project.project"
+
+    pr_state_default = fields.Selection(
+        selection=lambda self: self.env['project.task']._selection_pr_state(),
```
The better solution is to add a function for preparing selection values.
Function name: `_default_pr_state_default`.

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+
+    @api.model
+    def create(self, vals):
+        if 'project_id' in vals:
```
Better implementation:
```if 'project_id' in vals and 'pr_uri' in vals:```

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+        else:
+            vals['pr_state'] = None
+        task = super(ProjectTaskState, self).create(vals)
+        return task
```
No need one more create variable.
`return super(ProjectTaskState, self).create(vals)`

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+
+    def write(self, vals):
+        self._check_pr_state(vals)
+        result = super(ProjectTaskState, self).write(vals)
```
No need one more create variable.
return super(ProjectTaskState, self).create(vals)

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+        return result
+
+    def _check_pr_state(self, vals):
+        if 'pr_uri' in vals:
```
Better this condition take it to a `write` function.

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+        result = super(ProjectTaskState, self).write(vals)
+        return result
+
+    def _check_pr_state(self, vals):
```
This function makes pr_state values updated, but the function name is _check. Why?
Maybe the better solution will be renamed function and return new `pr_state` value?

geomer198 commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,47 @@
+        if 'pr_uri' in vals:
+            if vals['pr_uri']:
+                for rec in self:
+                    vals.update(pr_state=rec.project_id.pr_state_default)
```
`Vals` has dict type. This just updating one pr_state value in the cycle?

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,41 @@
+
+    @api.model
+    def create(self, vals):
+        if 'project_id' in vals and 'pr_uri' in vals:
```
Why aren't you using the same function for both write and create?
It will simplify the flow in case it is needed to be extended later.
![image](https://user-images.githubusercontent.com/2234999/215358976-eba6917d-1fe2-4bc6-b127-1cf0cf7f5e82.png)

ivs-cetmix commented on project_task_pull_request_state/models/project_task.py:
```diff
@@ -0,0 +1,41 @@
+        return super(ProjectTaskState, self).write(vals)
+
+    def _check_pr_state(self, vals):
+        if 'project_id' in vals:
```
Why are you checking for "project_id" here? What if only "pr_uri" is updated during write?

ivs-cetmix committed:
[IMP] project_task_pull_request_state

Fix _set_pr_state function. Set PR state automatically based on
project settings.
Tests are updated and now cover all possible cases.
Task View: Added pre-defined filter "Has PR" and group by PR state.

ivs-cetmix committed:
[DEL] project_task_pull_request_stare

Move to cetmix/cetmix-project repo
Will be brought back to OCA/project for 16.0
